### PR TITLE
Remove the "base_template_class" option on Twig\Environment

### PR DESCRIPTION
**Pull request by fabpot**

### CODE REVIEW

File: doc/internals.rst [hunk 1]
```diff
@@ -118,7 +118,7 @@ The generated template for a ``Hello {{ name }}`` template reads as follows
 using)::
 
     /* Hello {{ name }} */
-    class __TwigTemplate_1121b6f109fe93ebe8c6e22e3712bceb extends \Twig\Template
+    class __TwigTemplate_1121b6f109fe93ebe8c6e22e3712bceb extends Template
```
Comment (stof) [inline doc/internals.rst:121]: shouldn't we add the `use Twig\Template` in the code snippet then ?

Comment (fabpot) [reply doc/internals.rst:121]: Not really as we don't list the other use statements. It is just for illustration purpose.

Comment (stof) [reply doc/internals.rst:121]: but other used classes are not used in this snippet.

Comment (fabpot) [reply doc/internals.rst:121]: sure, but I don't see how this would improve the docs. I'm trying to keep the snippet short here.

**Review by stof — APPROVED** [review]

**Comment by hason** [comment]
@fabpot Was there any reason to remove this feature? I used a custom template class to load templates with a relative path to the currently rendered template (e.g. `{{ include('./parts/header.html.twig') }}`). How to solve this use case without this functionality?

**Comment by stof** [comment]
@hason can you share an example ?

Regarding the removal, this allowed to simplify the Twig codebase (and making the `Twig\Template` class internal, which unlocked some refactoring opportunities)

**Comment by hason** [comment]
@stof

```php
<?php

namespace App\Templating\Twig;

use Symfony\Component\Filesystem\Path;
use Twig\Template as BaseTemplate;

abstract class Template extends BaseTemplate
{
    protected function loadTemplate($template, $templateName = null, $line = null, $index = null)
    {
        $currentTemplate = $templateName ?? $this->getTemplateName();
        if (is_array($template)) {
            $template = array_map(fn ($template) => $this->resolveTemplate($currentTemplate, $template), $template);
        } else {
            $template = $this->resolveTemplate($currentTemplate, $template);
        }

        return parent::loadTemplate($template, $templateName, $line, $index);
    }

    private function resolveTemplate($currentTemplate, $template)
    {
        if (is_string($template) && str_starts_with($template, '.')) {
            return Path::canonicalize(Path::getDirectory($currentTemplate).'/'.$template);
        }

        return $template;
    }
}
```

**Comment by Terrijoo** [comment]
> @stof
> 
> ```
> <?php
> 
> namespace App\Templating\Twig;
> 
> use Symfony\Component\Filesystem\Path;
> use Twig\Template as BaseTemplate;
> 
> abstract class Template extends BaseTemplate
> {
>     protected function loadTemplate($template, $templateName = null, $line = null, $index = null)
>     {
>         $currentTemplate = $templateName ?? $this->getTemplateName();
>         if (is_array($template)) {
>             $template = array_map(fn ($template) => $this->resolveTemplate($currentTemplate, $template), $template);
>         } else {
>             $template = $this->resolveTemplate($currentTemplate, $template);
>         }
> 
>         return parent::loadTemplate($template, $templateName, $line, $index);
>     }
> 
>     private function resolveTemplate($currentTemplate, $template)
>     {
>         if (is_string($template) && str_starts_with($template, '.')) {
>             return Path::canonicalize(Path::getDirectory($currentTemplate).'/'.$template);
>         }
> 
>         return $template;
>     }
> }
> ```

Hey hason, did you ever found a workaround for it? We're struggling with a similar issue and it currently prevents us from upgrading to v3 of Twig.

**Pull request merged**